src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -100,103 +78,120 @@ public override void Accept(OperationVisitor visitor)
+                        RefKind refMode = argumentRefKinds.IsDefaultOrEmpty ? RefKind.None : argumentRefKinds[argumentIndex];
 
                         if (refMode != RefKind.None)
                         {
-                            return new Argument(ArgumentKind.Positional, parameterOpt, argument);
+                            return new Argument(ArgumentKind.Explicit, parameter, argument);
                         }
 
                         if (argumentIndex >= parameters.Length - 1 &&
                             parameters.Length > 0 &&
                             parameters[parameters.Length - 1].IsParams &&
                             // An argument that is an array of the appropriate type is not a params argument.
                             (boundArguments.Length > argumentIndex + 1 ||
-                             ((object)argument.Type != null && // If argument type is null, we are in an error scenario and cannot tell if it is a param array, or not. 
-                              (argument.Type.TypeKind != TypeKind.Array ||
-                              !argument.Type.Equals(parameters[parameters.Length - 1].Type, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds)))))
+                             argument.Type.TypeKind != TypeKind.Array ||
+                             !argument.Type.Equals(parameters[parameters.Length - 1].Type, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds)))
                         {
                             return new Argument(ArgumentKind.ParamArray, parameters[parameters.Length - 1], CreateParamArray(parameters[parameters.Length - 1], boundArguments, argumentIndex, invocationSyntax));
                         }
                         else
                         {
-                            return new SimpleArgument(parameterOpt, argument);
+                            return new Argument(ArgumentKind.Explicit, parameter, argument);
                         }
                     }
 
-                    return new Argument(ArgumentKind.Named, parameterOpt, argument);
+                    // A named argument which is also the only argument for a params parameter
+                    // is a IArgument of ArgumentKind.ParamArray. e.g.
+                    //
+                    //     static void M1(string[] args)
+                    //     {            
+                    //         M2(array: 1, str: "");
+                    //     }                             
+                    //     static void M2(string str, params int[] array) { }
+
+                    if (parameter?.IsParams == true &&
+                       // An argument that is an array of the appropriate type is not a params argument.
+                       (argument.Type.TypeKind != TypeKind.Array ||
+                       !argument.Type.Equals(parameter.Type, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds)))
```

Comment:
these conditions are too complex to easily parse and understand.  Please break out into sub pieces to make clearer.   Or just have nested ifs.